lightning/src/routing/router.rs

```diff
@@ -0,0 +1,1186 @@
+
+	// We do a dest-to-source Dijkstra's sorting by each node's distance from the destination
+	// plus the minimum per-HTLC fee to get from it to another node (aka "shitty A*").
+	// TODO: There are a few tweaks we could do, including possibly pre-calculating more stuff
+	// to use as the A* heuristic beyond just the cost to get one node further than the current
+	// one.
+
+	let dummy_directional_info = DummyDirectionalChannelInfo { // used for first_hops routes
+		src_node_id: network.our_node_id.clone(),
+		cltv_expiry_delta: 0,
+		htlc_minimum_msat: 0,
+		fees: RoutingFees {
+			base_msat: 0,
+			proportional_millionths: 0,
+		}
+	};
+
+	let mut targets = BinaryHeap::new(); //TODO: Do we care about switching to eg Fibbonaci heap?
+	let mut dist = HashMap::with_capacity(network.nodes.len());
+
+	let mut first_hop_targets = HashMap::with_capacity(if first_hops.is_some() { first_hops.as_ref().unwrap().len() } else { 0 });
+	if let Some(hops) = first_hops {
+		for chan in hops {
+			let short_channel_id = chan.short_channel_id.expect("first_hops should be filled in with usable channels, not pending ones");
+			if chan.remote_network_id == *target {
+				return Ok(Route {
+					paths: vec![vec![RouteHop {
+						pubkey: chan.remote_network_id,
+						node_features: NodeFeatures::with_known_relevant_init_flags(&chan.counterparty_features),
+						short_channel_id,
+						channel_features: ChannelFeatures::with_known_relevant_init_flags(&chan.counterparty_features),
+						fee_msat: final_value_msat,
+						cltv_expiry_delta: final_cltv,
+					}]],
+				});
+			}
+			first_hop_targets.insert(chan.remote_network_id, (short_channel_id, chan.counterparty_features.clone()));
+		}
+		if first_hop_targets.is_empty() {
+			return Err(LightningError{err: "Cannot route when there are no outbound routes away from us", action: ErrorAction::IgnoreError});
+		}
+	}
+
+	macro_rules! add_entry {
+		// Adds entry which goes from the node pointed to by $directional_info to
+		// $dest_node_id over the channel with id $chan_id with fees described in
+		// $directional_info.
+		( $chan_id: expr, $dest_node_id: expr, $directional_info: expr, $chan_features: expr, $starting_fee_msat: expr ) => {
+			//TODO: Explore simply adding fee to hit htlc_minimum_msat
+			if $starting_fee_msat as u64 + final_value_msat >= $directional_info.htlc_minimum_msat {
+				let proportional_fee_millions = ($starting_fee_msat + final_value_msat).checked_mul($directional_info.fees.proportional_millionths as u64);
+				if let Some(new_fee) = proportional_fee_millions.and_then(|part| {
+						($directional_info.fees.base_msat as u64).checked_add(part / 1000000) })
+				{
+					let mut total_fee = $starting_fee_msat as u64;
+					let hm_entry = dist.entry(&$directional_info.src_node_id);
+					let old_entry = hm_entry.or_insert_with(|| {
+						let node = network.nodes.get(&$directional_info.src_node_id).unwrap();
+						let mut fee_base_msat = 0;
+						let mut fee_proportional_millionths = 0;
```

Comment:
We wanted to get rid of default values in node inbound fees, but this is the place about I left them.
I'm not sure I fully understand what's going on in this part of the code, so not sure this is correct.

Reply:
Right, sorry, lack of clarity on my end - I was dont like having default values in public interfaces (as Rust has a native Option<> for exactly this reason), but in this case were putting default values in a temporary map used for our routefinding, which is totally fine.

Reply:
Right, I guess that's how the code should be read.